server/rest/src/main/java/org/infinispan/rest/resources/CounterResource.java

```diff
@@ -30,84 +50,299 @@
  * @since 10.0
  */
 public class CounterResource implements ResourceHandler {
-   private final CounterManager counterManager;
+   private final InvocationHelper invocationHelper;
+   private final EmbeddedCounterManager counterManager;
 
-   public CounterResource(CounterManager counterManager) {
-      this.counterManager = counterManager;
+   public CounterResource(InvocationHelper invocationHelper) {
+      this.invocationHelper = invocationHelper;
+      this.counterManager = invocationHelper.getCounterManager();
    }
 
    @Override
    public Invocations getInvocations() {
       return new Invocations.Builder()
-            .invocation().methods(POST).path("/v2/counters/{counterName}").handleWith(this::addCounter)
+            // Lifecycle
+            .invocation().methods(POST).path("/v2/counters/{counterName}").handleWith(this::createCounter)
+            .invocation().methods(DELETE).path("/v2/counters/{counterName}").handleWith(this::deleteCounter)
+
+            // Config
+            .invocation().methods(GET).path("/v2/counters/{counterName}/config").handleWith(this::getConfig)
+
+            // Counter ops
             .invocation().methods(GET).path("/v2/counters/{counterName}").handleWith(this::getCounter)
-            .invocation().methods(DELETE).path("/v2/counters/{counterName}").handleWith(this::resetCounter)
+            .invocation().methods(GET).path("/v2/counters/{counterName}").withAction("reset").handleWith(this::resetCounter)
+
+            // Weak counter ops
+            .invocation().methods(GET).path("/v2/counters/{counterName}").withAction("increment").handleWith(this::incrementWeak)
+            .invocation().methods(GET).path("/v2/counters/{counterName}").withAction("decrement").handleWith(this::decrementWeak)
+            .invocation().methods(GET).path("/v2/counters/{counterName}").withAction("add").handleWith(this::addValueWeak)
+
+            // Strong counter ops
+            .invocation().methods(GET).path("/v2/counters/{counterName}").withAction("incrementAndGet").handleWith(this::incrementStrong)
```

Comment:
Also, I think we should have a GET operation on the counters root resource which returns a summary of all counters.

If counter names are unique across weak/strong implementations, then maybe we should also allow a query string to return ^ GET method to just return counters of type strong or weak.

i.e GET `/v2/counters?type=strong`.

Obviously if counter names are not unique, we will want something like:

GET `v2/counters/strong`

but that will need all of the defined URLs updated.

Reply:
> Also, I think we should have a GET operation on the counters root resource which returns a summary of all counters

Would it return a list of configurations? Or the current value of the counter?

Reply:
it is not possible. the counters name are unique independently of the configuration used.

Reply:
> i.e GET /v2/counters?type=strong

This is reasonable, but same question as before: would it return the name/config pairs, or name/value pairs?

Reply:
or maybe name/config/current value?

Reply:
@gustavonalle As the counters value will just be a long, how about all three name/config/value

Reply:
Could also parameterise the return type, e.g., don't want the configs since it can be obtained via other endopoints. Let me handle this in a separate JIRA: https://issues.jboss.org/browse/ISPN-10401